File: sdks/typescript/tests/errors.test.ts
/**
 * Tests for the error hierarchy.
 */

import {
  ReinforceSpecError,
  ValidationError,
  AuthenticationError,
  AuthorizationError,
  NotFoundError,
  ConflictError,
  RateLimitError,
  PayloadTooLargeError,
  ServerError,
  ServiceUnavailableError,
  NetworkError,
  TimeoutError,
  errorFromResponse,
} from '../src/errors';

describe('Error classes', () => {
  it('all errors should extend ReinforceSpecError', () => {
    const errors = [
      new ValidationError('test'),
      new AuthenticationError(),
      new AuthorizationError(),
      new NotFoundError(),
      new ConflictError(),
      new RateLimitError(),
      new PayloadTooLargeError(),
      new ServerError(),
      new ServiceUnavailableError(),
      new NetworkError(),
      new TimeoutError(),
    ];

    for (const err of errors) {
      expect(err).toBeInstanceOf(ReinforceSpecError);
      expect(err).toBeInstanceOf(Error);
    }
  });

  it('ReinforceSpecError should set properties', () => {
    const err = new ReinforceSpecError('test message', {
      statusCode: 500,
      details: { key: 'value' },
    });
    expect(err.message).toBe('test message');
    expect(err.statusCode).toBe(500);
    expect(err.details).toEqual({ key: 'value' });
  });

  it('ValidationError should set field and value', () => {
    const err = new ValidationError('bad input', {
      field: 'candidates',
      value: [],
    });
    expect(err.field).toBe('candidates');
    expect(err.value).toEqual([]);
    expect(err.statusCode).toBe(400);
  });

  it('RateLimitError should set retry info', () => {
    const err = new RateLimitError('rate limited', {
      retryAfter: 30000,
      limit: 100,
      remaining: 0,
      resetAt: 1700000000,
    });
    expect(err.retryAfter).toBe(30000);
    expect(err.limit).toBe(100);
    expect(err.remaining).toBe(0);
    expect(err.statusCode).toBe(429);
  });

  it('NotFoundError should set resource info', () => {
    const err = new NotFoundError('not found', {
      resourceType: 'policy',
      resourceId: 'v001',
    });
    expect(err.resourceType).toBe('policy');
    expect(err.resourceId).toBe('v001');
    expect(err.statusCode).toBe(404);
  });

  it('TimeoutError should set timeout', () => {
    const err = new TimeoutError('timed out', 30000);
    expect(err.timeout).toBe(30000);
  });

  it('NetworkError should set cause', () => {
    const cause = new TypeError('fetch failed');
    const err = new NetworkError('connection failed', cause);
    expect(err.cause).toBe(cause);
  });
});

describe('errorFromResponse', () => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const testCases: Array<[number, any]> = [
    [400, ValidationError],
    [401, AuthenticationError],
    [403, AuthorizationError],
    [404, NotFoundError],
    [409, ConflictError],
    [413, PayloadTooLargeError],
    [429, RateLimitError],
    [500, ServerError],
    [502, ServerError],
    [503, ServiceUnavailableError],
    [504, ServerError],
  ];

  it.each(testCases)('status %i should produce %s', (status, ErrorClass) => {
    const err = errorFromResponse(status, 'test error');
    expect(err).toBeInstanceOf(ErrorClass);
  });

  it('unknown status should produce base error', () => {
    const err = errorFromResponse(418, "I'm a teapot");
    expect(err).toBeInstanceOf(ReinforceSpecError);
    expect(err.statusCode).toBe(418);
  });

  it('429 should extract rate limit headers', () => {
    const headers = new Headers({
      'Retry-After': '30',
      'X-RateLimit-Limit': '100',
      'X-RateLimit-Remaining': '0',
    });
    const err = errorFromResponse(429, 'rate limited', undefined, headers);
    expect(err).toBeInstanceOf(RateLimitError);
    const rle = err as RateLimitError;
    expect(rle.retryAfter).toBe(30000);
    expect(rle.limit).toBe(100);
    expect(rle.remaining).toBe(0);
  });
});

File: sdks/go/errors_test.go
package reinforcespec

import (
	"errors"
	"testing"
	"time"
)

func TestErrorHierarchy(t *testing.T) {
	tests := []struct {
		name     string
		err      error
		sentinel error
	}{
		{"ValidationError", &ValidationError{ReinforceSpecError: ReinforceSpecError{Message: "bad"}}, ErrValidation},
		{"AuthenticationError", &AuthenticationError{ReinforceSpecError: ReinforceSpecError{Message: "unauthorized"}}, ErrAuthentication},
		{"AuthorizationError", &AuthorizationError{ReinforceSpecError: ReinforceSpecError{Message: "forbidden"}}, ErrAuthorization},
		{"NotFoundError", &NotFoundError{ReinforceSpecError: ReinforceSpecError{Message: "missing"}}, ErrNotFound},
		{"ConflictError", &ConflictError{ReinforceSpecError: ReinforceSpecError{Message: "conflict"}}, ErrConflict},
		{"RateLimitError", &RateLimitError{ReinforceSpecError: ReinforceSpecError{Message: "throttled"}}, ErrRateLimit},
		{"ServerError", &ServerError{ReinforceSpecError: ReinforceSpecError{Message: "internal"}}, ErrServer},
		{"NetworkError", &NetworkError{ReinforceSpecError: ReinforceSpecError{Message: "conn"}}, ErrNetwork},
		{"TimeoutError", &TimeoutError{ReinforceSpecError: ReinforceSpecError{Message: "timeout"}}, ErrTimeout},
		{"ConfigurationError", &ConfigurationError{ReinforceSpecError: ReinforceSpecError{Message: "config"}}, ErrConfiguration},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if !errors.Is(tt.err, tt.sentinel) {
				t.Errorf("%s should wrap %v", tt.name, tt.sentinel)
			}
		})
	}
}

func TestErrorFromResponse(t *testing.T) {
	tests := []struct {
		status   int
		expected string
	}{
		{400, "*reinforcespec.ValidationError"},
		{401, "*reinforcespec.AuthenticationError"},
		{403, "*reinforcespec.AuthorizationError"},
		{404, "*reinforcespec.NotFoundError"},
		{409, "*reinforcespec.ConflictError"},
		{429, "*reinforcespec.RateLimitError"},
		{500, "*reinforcespec.ServerError"},
		{502, "*reinforcespec.ServerError"},
		{503, "*reinforcespec.ServerError"},
		{504, "*reinforcespec.ServerError"},
	}

	for _, tt := range tests {
		t.Run(tt.expected, func(t *testing.T) {
			err := errorFromResponse(tt.status, "test", nil)
			if err == nil {
				t.Fatal("expected error, got nil")
			}
		})
	}
}

func TestErrorFromResponseRateLimitDetails(t *testing.T) {
	details := map[string]interface{}{
		"retry_after": float64(30),
		"limit":       float64(100),
		"remaining":   float64(0),
	}
	err := errorFromResponse(429, "rate limited", details)

	rle, ok := err.(*RateLimitError)
	if !ok {
		t.Fatalf("expected *RateLimitError, got %T", err)
	}
	if rle.RetryAfter != 30*time.Second {
		t.Errorf("expected retry after 30s, got %v", rle.RetryAfter)
	}
	if rle.Limit != 100 {
		t.Errorf("expected limit 100, got %d", rle.Limit)
	}
}

func TestReinforceSpecErrorFormat(t *testing.T) {
	err := &ReinforceSpecError{
		Message:    "test error",
		StatusCode: 500,
	}
	want := "[500] test error"
	if got := err.Error(); got != want {
		t.Errorf("got %q, want %q", got, want)
	}
}

func TestReinforceSpecErrorFormatNoStatus(t *testing.T) {
	err := &ReinforceSpecError{Message: "test error"}
	if got := err.Error(); got != "test error" {
		t.Errorf("got %q, want %q", got, "test error")
	}
}

func TestValidationErrorFormat(t *testing.T) {
	err := &ValidationError{
		ReinforceSpecError: ReinforceSpecError{Message: "bad input"},
		Field:              "candidates",
	}
	got := err.Error()
	if got != "validation error: bad input (field: candidates)" {
		t.Errorf("got %q", got)
	}
}

func TestNotFoundErrorFormat(t *testing.T) {
	err := &NotFoundError{
		ReinforceSpecError: ReinforceSpecError{Message: "missing"},
		ResourceType:       "policy",
		ResourceID:         "v001",
	}
	got := err.Error()
	if got != "policy not found: v001" {
		t.Errorf("got %q", got)
	}
}
